### ISSUE TITLE
Install and proxy conflict.

### ISSUE THREAD

**Issue by KyleFontenot** [lovell/sharp#3441]
Given these issues: [#1845,](https://github.com/lovell/sharp/issues/1845), [#2019](https://github.com/lovell/sharp/issues/2019), [#2950](https://github.com/lovell/sharp/issues/2950), and [#3051](https://github.com/lovell/sharp/issues/3051)
I still have problems installing sharp when using a proxy.

Proxy settings are set for local Linux OS, npm, yarn, and git. Npm and git work perfectly fine. I can install all packages, no problem, except for sharp.
Installing sharp crashes when executing the `(node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)` command (as shown in my full error), because it crashes here:
```
sharp: Downloading https://github.com/lovell/sharp-libvips/releases/download/v8.13.3/libvips-8.13.3-linux-x64.tar.brnstall/
sharp: Installation error: tunneling socket could not be established, cause=write EPROTO C0473B72537F0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:355:
```
How do I get sharp's install script to use my proxy settings?

## Possible install-time or require-time problem

- [x] I have read the [documentation relating to installation](https://sharp.pixelplumbing.com/install).
- [x] I have ensured that the architecture and platform of Node.js used for `npm install` is the same as the architecture and platform of Node.js used at runtime.

### Are you using the latest version of sharp?

- [x] I am using the latest version of `sharp` as reported by `npm view sharp dist-tags.latest`.

### Is this a problem with filesystem permissions?

If you are using npm v6 or earlier and installing as a `root` or `sudo` user, have you tried with the `npm install --unsafe-perm` flag?

- I'm using npm 8.19.2, I have still tried --unsafe-perm anyways.

If you are using npm v7 or later, does the user running `npm install` own the directory it is run in?
- Yes it does. Quadruple-checked

If you are using the `ignore-scripts` feature of `npm`, have you tried with the `npm install --ignore-scripts=false` flag?
- N/A

### What is the complete output of running `npm install --verbose --foreground-scripts sharp` in an empty directory?

<details>

npm verb cli /home/radmed/.nvm/versions/node/v18.12.1/bin/node /home/radmed/.nvm/versions/node/v18.12.1/bin/npm 
npm info using npm@8.19.2
npm info using node@v18.12.1
npm timing npm:load:whichnode Completed in 1ms
npm timing config:load:defaults Completed in 3ms
npm timing config:load:file:/home/radmed/.nvm/versions/node/v18.12.1/lib/node_modules/npm/npmrc Completed in 1ms
npm timing config:load:builtin Completed in 2ms
npm timing config:load:cli Completed in 4ms
npm timing config:load:env Completed in 0ms
npm timing config:load:file:/var/www/testdir/.npmrc Completed in 0ms
npm timing config:load:project Completed in 7ms
npm timing config:load:file:/home/radmed/.npmrc Completed in 3ms
npm timing config:load:user Completed in 3ms
npm timing config:load:file:/home/radmed/.nvm/versions/node/v18.12.1/etc/npmrc Completed in 0ms
npm timing config:load:global Completed in 0ms
npm timing config:load:validate Completed in 2ms
npm timing config:load:credentials Completed in 2ms
npm timing config:load:setEnvs Completed in 3ms
npm timing config:load Completed in 28ms
npm timing npm:load:configload Completed in 28ms
npm timing npm:load:mkdirpcache Completed in 8ms
npm timing npm:load:mkdirplogs Completed in 1ms
npm verb title npm install sharp
npm verb argv "install" "--loglevel" "verbose" "--foreground-scripts" "sharp"
npm timing npm:load:setTitle Completed in 4ms
npm timing config:load:flatten Completed in 6ms
npm timing npm:load:display Completed in 16ms
npm verb logfile logs-max:10 dir:/home/radmed/.npm/_logs
npm verb logfile /home/radmed/.npm/_logs/2022-11-08T16_55_49_915Z-debug-0.log
npm timing npm:load:logFile Completed in 10ms
npm timing npm:load:timers Completed in 0ms
npm timing npm:load:configScope Completed in 0ms
npm timing npm:load Completed in 69ms
npm timing arborist:ctor Completed in 2ms
npm timing idealTree:init Completed in 15ms
npm timing idealTree:userRequests Completed in 4ms
npm http fetch GET 200 https://registry.npmjs.org/sharp 443ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/tar-fs 129ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/node-addon-api 147ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/simple-get 158ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/detect-libc 177ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/color 204ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/prebuild-install 363ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/tunnel-agent 368ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/semver 379ms (cache revalidated)
npm timing idealTree:#root Completed in 869ms
npm http fetch GET 200 https://registry.npmjs.org/minimist 211ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/once 213ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/lru-cache 216ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/expand-template 238ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/tar-stream 217ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/node-abi 250ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/rc 315ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/pump 387ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/color-convert 423ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/mkdirp-classic 418ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/chownr 439ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/github-from-package 461ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/color-string 595ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/simple-concat 593ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/mkdirp-classic 635ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/napi-build-utils 640ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/safe-buffer 804ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/decompress-response 1450ms (cache revalidated)
npm timing idealTree:node_modules/sharp Completed in 1512ms
npm http fetch GET 200 https://registry.npmjs.org/color-name 105ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/simple-swizzle 305ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/color-name 340ms (cache revalidated)
npm timing idealTree:node_modules/color Completed in 355ms
npm timing idealTree:node_modules/color-convert Completed in 4ms
npm timing idealTree:node_modules/color-name Completed in 0ms
npm http fetch GET 200 https://registry.npmjs.org/is-arrayish 78ms (cache revalidated)
npm timing idealTree:node_modules/color-string Completed in 85ms
npm timing idealTree:node_modules/detect-libc Completed in 0ms
npm timing idealTree:node_modules/node-addon-api Completed in 0ms
npm http fetch GET 200 https://registry.npmjs.org/deep-extend 91ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/ini 128ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/end-of-stream 141ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/strip-json-comments 305ms (cache revalidated)
npm timing idealTree:node_modules/prebuild-install Completed in 352ms
npm timing idealTree:node_modules/expand-template Completed in 0ms
npm timing idealTree:node_modules/github-from-package Completed in 0ms
npm timing idealTree:node_modules/minimist Completed in 0ms
npm timing idealTree:node_modules/mkdirp-classic Completed in 0ms
npm timing idealTree:node_modules/napi-build-utils Completed in 0ms
npm timing idealTree:node_modules/node-abi Completed in 0ms
npm http fetch GET 200 https://registry.npmjs.org/wrappy 307ms (cache revalidated)
npm timing idealTree:node_modules/pump Completed in 320ms
npm timing idealTree:node_modules/end-of-stream Completed in 0ms
npm timing idealTree:node_modules/once Completed in 3ms
npm timing idealTree:node_modules/rc Completed in 11ms
npm timing idealTree:node_modules/deep-extend Completed in 0ms
npm timing idealTree:node_modules/ini Completed in 0ms
npm http fetch GET 200 https://registry.npmjs.org/yallist 114ms (cache revalidated)
npm timing idealTree:node_modules/semver Completed in 122ms
npm timing idealTree:node_modules/lru-cache Completed in 4ms
npm http fetch GET 200 https://registry.npmjs.org/mimic-response 310ms (cache revalidated)
npm timing idealTree:node_modules/simple-get Completed in 323ms
npm timing idealTree:node_modules/decompress-response Completed in 3ms
npm timing idealTree:node_modules/mimic-response Completed in 0ms
npm timing idealTree:node_modules/simple-concat Completed in 0ms
npm timing idealTree:node_modules/simple-swizzle Completed in 3ms
npm timing idealTree:node_modules/is-arrayish Completed in 0ms
npm timing idealTree:node_modules/strip-json-comments Completed in 0ms
npm http fetch GET 200 https://registry.npmjs.org/bl 129ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/readable-stream 128ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/fs-constants 145ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/inherits 154ms (cache revalidated)
npm timing idealTree:node_modules/tar-fs Completed in 169ms
npm timing idealTree:node_modules/chownr Completed in 0ms
npm http fetch GET 200 https://registry.npmjs.org/string_decoder 118ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/util-deprecate 271ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/buffer 375ms (cache revalidated)
npm timing idealTree:node_modules/tar-stream Completed in 398ms
npm http fetch GET 200 https://registry.npmjs.org/base64-js 168ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/ieee754 561ms (cache revalidated)
npm timing idealTree:node_modules/bl Completed in 571ms
npm timing idealTree:node_modules/buffer Completed in 7ms
npm timing idealTree:node_modules/base64-js Completed in 0ms
npm timing idealTree:node_modules/fs-constants Completed in 0ms
npm timing idealTree:node_modules/ieee754 Completed in 0ms
npm timing idealTree:node_modules/inherits Completed in 0ms
npm timing idealTree:node_modules/readable-stream Completed in 8ms
npm timing idealTree:node_modules/string_decoder Completed in 4ms
npm timing idealTree:node_modules/safe-buffer Completed in 0ms
npm timing idealTree:node_modules/tunnel-agent Completed in 0ms
npm timing idealTree:node_modules/util-deprecate Completed in 0ms
npm timing idealTree:node_modules/wrappy Completed in 0ms
npm timing idealTree:node_modules/yallist Completed in 0ms
npm timing idealTree:buildDeps Completed in 5145ms
npm timing idealTree:fixDepFlags Completed in 4ms
npm timing idealTree Completed in 5173ms
npm timing reify:loadTrees Completed in 5174ms
npm timing reify:diffTrees Completed in 6ms
npm timing reify:retireShallow Completed in 4ms
npm timing reify:createSparse Completed in 28ms
npm timing reify:loadBundles Completed in 0ms
npm timing reifyNode:node_modules/chownr Completed in 242ms
npm timing reifyNode:node_modules/strip-json-comments Completed in 242ms
npm timing reifyNode:node_modules/ini Completed in 243ms
npm timing reifyNode:node_modules/wrappy Completed in 242ms
npm timing reifyNode:node_modules/once Completed in 243ms
npm timing reifyNode:node_modules/end-of-stream Completed in 243ms
npm timing reifyNode:node_modules/mkdirp-classic Completed in 243ms
npm timing reifyNode:node_modules/simple-swizzle Completed in 244ms
npm timing reifyNode:node_modules/color-name Completed in 245ms
npm timing reifyNode:node_modules/color-string Completed in 244ms
npm timing reifyNode:node_modules/inherits Completed in 251ms
npm timing reifyNode:node_modules/ieee754 Completed in 253ms
npm timing reifyNode:node_modules/fs-constants Completed in 252ms
npm timing reifyNode:node_modules/mimic-response Completed in 252ms
npm timing reifyNode:node_modules/lru-cache Completed in 252ms
npm timing reifyNode:node_modules/decompress-response Completed in 252ms
npm timing reifyNode:node_modules/tunnel-agent Completed in 251ms
npm timing reifyNode:node_modules/color Completed in 251ms
npm timing reifyNode:node_modules/string_decoder Completed in 258ms
npm timing reifyNode:node_modules/util-deprecate Completed in 260ms
npm timing reifyNode:node_modules/base64-js Completed in 259ms
npm timing reifyNode:node_modules/yallist Completed in 257ms
npm timing reifyNode:node_modules/expand-template Completed in 256ms
npm timing reifyNode:node_modules/simple-concat Completed in 261ms
npm timing reifyNode:node_modules/deep-extend Completed in 260ms
npm timing reifyNode:node_modules/pump Completed in 261ms
npm timing reifyNode:node_modules/napi-build-utils Completed in 260ms
npm timing reifyNode:node_modules/safe-buffer Completed in 268ms
npm timing reifyNode:node_modules/is-arrayish Completed in 266ms
npm timing reifyNode:node_modules/color-convert Completed in 265ms
npm timing reifyNode:node_modules/detect-libc Completed in 265ms
npm timing reifyNode:node_modules/tar-stream Completed in 273ms
npm timing reifyNode:node_modules/simple-get Completed in 272ms
npm timing reifyNode:node_modules/buffer Completed in 279ms
npm timing reifyNode:node_modules/github-from-package Completed in 277ms
npm timing reifyNode:node_modules/rc Completed in 279ms
npm timing reifyNode:node_modules/bl Completed in 283ms
npm timing reifyNode:node_modules/prebuild-install Completed in 280ms
npm timing reifyNode:node_modules/node-abi Completed in 284ms
npm http fetch POST 200 https://registry.npmjs.org/-/npm/v1/security/advisories/bulk 309ms
npm timing auditReport:getReport Completed in 319ms
npm timing auditReport:init Completed in 0ms
npm timing reify:audit Completed in 321ms
npm timing reifyNode:node_modules/minimist Completed in 305ms
npm timing reifyNode:node_modules/readable-stream Completed in 317ms
npm timing reifyNode:node_modules/node-addon-api Completed in 317ms
npm timing reifyNode:node_modules/tar-fs Completed in 319ms
npm timing reifyNode:node_modules/semver Completed in 323ms
npm timing reifyNode:node_modules/sharp Completed in 327ms
npm timing reify:unpack Completed in 336ms
npm timing reify:unretire Completed in 0ms
npm timing build:queue Completed in 5ms
npm timing build:link:node_modules/semver Completed in 3ms
npm timing build:link:node_modules/prebuild-install Completed in 5ms
npm timing build:link:node_modules/rc Completed in 4ms
npm timing build:link Completed in 6ms
npm info run sharp@0.31.2 install node_modules/sharp (node install/libvips && node install/dll-copy && prebuild-install) || 
(node install/can-compile && node-gyp rebuild && node install/dll-copy)

> sharp@0.31.2 install
> (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)

sharp: Downloading https://github.com/lovell/sharp-libvips/releases/download/v8.13.3/libvips-8.13.3-linux-x64.tar.brnstall/ 
sharp: Installation error: tunneling socket could not be established, cause=write EPROTO C0473B72537F0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:355:

sharp: Please see https://sharp.pixelplumbing.com/install for required dependencies
npm info run sharp@0.31.2 install { code: 1, signal: null }
npm timing reify:rollback:createSparse Completed in 101ms
npm timing reify:rollback:retireShallow Completed in 0ms
npm timing command:install Completed in 6176ms
npm verb stack Error: command failed
npm verb stack     at ChildProcess.<anonymous> (/home/radmed/.nvm/versions/node/v18.12.1/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/lib/index.js:63:27)
npm verb stack     at ChildProcess.emit (node:events:513:28)
npm verb stack     at maybeClose (node:internal/child_process:1091:16)
npm verb stack     at ChildProcess._handle.onexit (node:internal/child_process:302:5)
npm verb pkgid sharp@0.31.2
npm verb cwd /var/www/testdir
npm verb Linux 5.15.0-52-generic
npm verb node v18.12.1
npm verb npm  v8.19.2
npm ERR! code 1
npm ERR! path /var/www/testdir/node_modules/sharp
npm ERR! command failed
npm ERR! command sh -c -- (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)
npm verb exit 1
npm timing npm Completed in 6414ms
npm verb unfinished npm timer reify 1667926550148
npm verb unfinished npm timer reify:build 1667926555737
npm verb unfinished npm timer build 1667926555737
npm verb unfinished npm timer build:deps 1667926555738
npm verb unfinished npm timer build:run:install 1667926555750
npm verb unfinished npm timer build:run:install:node_modules/sharp 1667926555751
npm verb code 1

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/radmed/.npm/_logs/2022-11-08T16_55_49_915Z-debug-0.log
</details>

### What is the output of running `npx envinfo --binaries --system --npmPackages=sharp --npmGlobalPackages=sharp`?

```
System:
    OS: Linux 5.15 Ubuntu 22.04.1 LTS 22.04.1 LTS (Jammy Jellyfish)
    CPU: (24) x64 Intel(R) Xeon(R) CPU E5-2620 0 @ 2.00GHz
    Memory: 17.76 GB / 31.32 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v18.12.1/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
```

**Comment by lovell** [comment]
> sharp: Installation error: tunneling socket could not be established, cause=write EPROTO C0473B72537F0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:355:

How are you configuring the proxy? Does it support SSL/TLS connections? Are you sure the protocol (http vs https) is correct?

**Comment by lovell** [comment]
My best guess would be that the proxy service running on port 8080 doesn't support encryption. If so, this should be configured as `HTTPS_PROXY="http://<proxyIPaddr>:8080"` (note `http` rather than `https`).

As an aside, if this is the case, then traffic between your machine and the proxy is not encrypted so please be careful.

**Comment by KyleFontenot** [comment]
I've set `HTTPS_PROXY` to `"http://<proxyIPaddr>:8080"` instead (no s). 
The problem remains

**Comment by lovell** [comment]
The original error looks like a proxy pretending to support TLS but not really doing so. The other settings suggest the proxy you are using does not support TLS.

Are you sure that all configured proxy values start with `http://...`? Are there any proxy-related environment variables or npm settings not shown here?

**Comment by KyleFontenot** [comment]
No other environment variables anywhere else; it was a clean install from Monday, so I know exactly where I set them. And I can't think of any other conf file that would need proxy set, except maybe apt.conf for Ubuntu, but I definitely do not have that set.

But, trying it again today, it worked. 
Literally the only thing that changed is that I restarted my local computer (my windows box that I'm SSHing into the Ubuntu server with). 
So the problem magically disappeared as magically as it appeared. 

I'll open this back up if it pops up again.
